KachnaOnline.Business/Services/UserService.cs

```diff
@@ -277,5 +283,74 @@ private async Task MapRoles(UserEntity userEntity, string[] userKisRoles)
             foreach (var userRole in toRemove)
                 userEntity.Roles.Remove(userRole);
         }
+
+        /// <inheritdoc />
+        public async Task<IEnumerable<Role>> GetRoles()
+        {
+            var roles = await _roleRepository.All().ToListAsync();
+            return _mapper.Map<IEnumerable<Role>>(roles);
+        }
+
+        /// <inheritdoc />
+        public async Task<Role> GetRole(int id)
+        {
+            var role = await _roleRepository.Get(id);
+            if (role is null)
+            {
+                throw new RoleNotFoundException(id);
+            }
+
+            return _mapper.Map<Role>(role);
+        }
+
+        /// <inheritdoc />
+        public async Task<IEnumerable<User>> GetUsers()
+        {
+            var users = await _userRepository.All().ToListAsync();
+            return _mapper.Map<IEnumerable<User>>(users);
+        }
+
+        /// <inheritdoc />
+        public async Task AssignRole(UserRole assignment)
+        {
+            if (await _userRoleRepository.Get(assignment.UserId, assignment.RoleId) is not null)
+            {
+                throw new RoleAlreadyAssignedException(assignment.UserId, assignment.RoleId);
+            }
+
+            await _userRoleRepository.Add(_mapper.Map<UserRoleEntity>(assignment));
+            try
+            {
+                await _unitOfWork.SaveChanges();
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e, "Role assignment failed.");
+                await _unitOfWork.ClearTrackedChanges();
+                throw new RoleManipulationFailedException();
+            }
+        }
+
+        /// <inheritdoc />
+        public async Task RevokeRole(int userId, int roleId)
+        {
+            var entity = await _userRoleRepository.Get(userId, roleId);
+            if (entity is null)
+            {
+                throw new RoleNotFoundException();
+            }
+
+            await _userRoleRepository.Delete(entity);
+            try
+            {
+                await _unitOfWork.SaveChanges();
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e, "Role revocation failed.");
+                await _unitOfWork.ClearTrackedChanges();
+                throw new RoleManipulationFailedException();
+            }
+        }
```

Comment:
This is wrong. The role system is designed around mapping KIS roles to local roles. Deleting a `UserRole` entry doesn't do anything useful if the role has been mapped from KIS. It will be recreated when the user next logs in. That's why it has the `ForceDisable` and `AssignedByUser` properties. 

Basically, there shouldn't even be need for explicitly changing one's roles. If a manual change of roles is performed, there probably is a very good reason for it, it means that the local user should have a different scope of permissions than they would normally have with their KIS role; and it has to be explicitly noted in the role assignment. Looking at it, my original `RoleAssignment` model isn't designed this way and an `IsDisabled` property should be added.

When a role is *manually* assigned, the `AssignedByUserId` attribute of the `UserRole` entity **must** be set. A non-null value of this field signalizes (to the role mapper) that it is a manually assigned role. This ensures that the role won't be *dropped* when the user next logs in (because they don't have the corresponding KIS role).

When a role is manually revoked, the entity **must not** be deleted (**if** it hasn't been created manually before); instead, `AssignedByUserId` must be set and `ForceDisable` must be set to `true`. This ensures that the mapping won't be recreated based on the user's KIS role.

This should **not** be opaque to the user. Like I've said, such changes are non-standard and the administrator should clearly see whether this role assignment is or is not KIS-mapped. There should be two operations available: *revert* role assignment, that actually drops the entity and reverts the beahivour back to default, that is, KIS-mapped; and *revoke* role from the user, that performs the operations mentioned above.